### PR TITLE
Make message streams accept XML and JSON data

### PR DESCRIPTION
**Pull request by peterschrammel**
and other refactorings towards replacing cout.
More to come to replace cout in show_properties show_goto_functions, etc

### CODE REVIEW

**Review by smowton — APPROVED** [review]

File: src/cbmc/bmc_cover.cpp [hunk 1]
```diff
@@ -280,26 +272,25 @@ bool bmc_covert::operator()()
...
-        status() << "[" << g.first << "]";
+        result() << "[" << g.first << "]";
 
         if(goal.source_location.is_not_nil())
-          status() << ' ' << goal.source_location;
+          result() << ' ' << goal.source_location;
 
-        if(!goal.description.empty())
-          status() << ' ' << goal.description;
+        if(goal.description.empty())
```
Comment (lucasccordeiro) [inline src/cbmc/bmc_cover.cpp]: The condition of the if-statement was changed. Is that correct?

**Review by lucasccordeiro — APPROVED** [review]

File: src/analyses/natural_loops.h [hunk 2]
```diff
@@ -15,6 +15,7 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
 #include <stack>
 #include <iosfwd>
 #include <set>
+#include <ostream>
```
Comment (kroening) [inline src/analyses/natural_loops.h]: Sure the ostream include is needed? This is a heavy include.

File: src/cbmc/all_properties.cpp [hunk 3]
```diff
@@ -224,7 +224,7 @@ void bmc_all_propertiest::report(const cover_goalst &cover_goals)
         }
       }
 
-      std::cout << ",\n" << json_result;
+      result() << preformatted_output << json_result << eom;
     }
```
Comment (kroening) [inline src/cbmc/all_properties.cpp]: It doesn't make sense to send json to the message streams.

File: src/cbmc/bmc.cpp [hunk 4]
```diff
@@ -173,16 +173,15 @@ void bmct::report_success()
     {
       xmlt xml("cprover-status");
       xml.data="SUCCESS";
-      std::cout << xml;
-      std::cout << "\n";
+      result() << preformatted_output << xml << eom;
     }
```
Comment (kroening) [inline src/cbmc/bmc.cpp]: Same here -- no XML into the message streams. In the extreme case, the message stream might then produce, say, JSON, which will then have escaped XML in it.

File: src/cbmc/bmc_cover.cpp [hunk 5]
```diff
@@ -317,7 +308,7 @@ bool bmc_covert::operator()()
         if(goal.source_location.is_not_nil())
           xml_result.new_element()=xml(goal.source_location);
 
-        std::cout << xml_result << "\n";
+        result() << preformatted_output << xml_result << '\n';
       }
```
Comment (kroening) [inline src/cbmc/bmc_cover.cpp]: Same here.

**Comment by kroening** [comment]
The key problem here is that JSON/XML gets put into the message streams, which themselves might encode things in XML or JSON.

We can contemplate a raw() stream if you want to avoid std::cout.

**Comment by tautschnig** [comment]
Just noting that #1339 and this one will need to be synchronised carefully, far as I am aware.

**Comment by peterschrammel** [comment]
The way XML and JSON is sent to the output is a hack (in particular for JSON (blaming myself) where delimiters are required, see https://github.com/diffblue/cbmc/pull/1349#discussion-diff-137219418L227).

`preformatted_output` was introduced to eliminate this problem and provide a mechanism to correctly intersperse formatted output with messages. It turns a message stream into a raw stream for the current message. 

We could replace `preformatted_output` by a `raw()` stream, if it is acceptable that verbosity levels need not be distinguished for raw outputs.

**Comment by kroening** [comment]
Any scope for giving the xml or the json to the message handler instead of the ascii version?

**Comment by peterschrammel** [comment]
Yes, that can be done, and would also catch when someone is trying to feed xml into a json-ui...

**Comment by kroening** [comment]
Ok, sounds good, let's go for it.

File: src/util/message.h [hunk 6]
```diff
@@ -176,6 +188,26 @@ class messaget
     messaget &message;
     source_locationt source_location;
 
+    mstreamt &operator << (const xmlt &data)
+    {
+      *this << eom; // force end of previous message
+      if(message.message_handler)
```
Comment (kroening) [inline src/util/message.h]: Is that really needed? Might produce spurious blank messages?

**Comment by kroening** [comment]
Clearly superior to the previous approach!

I'd squash the "override"s.

**Pull request merged**